File: app/src/main/java/com/team2052/frckrawler/database/tables/AbstractTable.java
package com.team2052.frckrawler.database.tables;

import com.team2052.frckrawler.database.DBManager;

import org.greenrobot.greendao.AbstractDao;
import org.greenrobot.greendao.query.QueryBuilder;

import java.util.List;

abstract class AbstractTable<T, D extends AbstractDao<T, Long>> {
    protected DBManager dbManager;
    D dao;

    public AbstractTable(D dao, DBManager dbManager) {
        this.dao = dao;
        this.dbManager = dbManager;
    }

    public abstract T load(long id);

    public abstract void delete(T model);

    public void delete(List<T> models) {
        for (int i = 0; i < models.size(); i++) {
            delete(models.get(i));
        }
    }

    public QueryBuilder<T> getQueryBuilder() {
        return dao.queryBuilder();
    }

    public D getDao() {
        return dao;
    }

    public List<T> loadAll() {
        return dao.loadAll();
    }

    public abstract void insert(T model);

    public void deleteAll() {
        dao.deleteAll();
    }
}

File: app/src/main/java/com/team2052/frckrawler/data/tables/AbstractTable.kt
package com.team2052.frckrawler.data.tables

import com.team2052.frckrawler.data.DBManager
import org.greenrobot.greendao.AbstractDao
import org.greenrobot.greendao.query.QueryBuilder

abstract class AbstractTable<T, D : AbstractDao<T, Long>>(var dao: D, protected var dbManager: DBManager) {
    open fun load(id: Long?): T? = dao.load(id)
    open fun delete(model: T) {
        dao.delete(model)
    }

    open fun insert(model: T): Long {
        return dao.insert(model)
    }

    fun delete(models: List<T>) {
        models.forEach { delete(it) }
    }

    fun loadAll(): List<T> {
        return dao.loadAll()
    }

    val queryBuilder: QueryBuilder<T>
        get() = dao.queryBuilder()

    fun deleteAll() {
        dao.deleteAll()
    }
}
